src/plot_api/plot_api.js

```diff
@@ -1576,9 +1576,17 @@ function _restyle(gd, aobj, _traces) {
                 helpers.manageArrayContainers(param, newVal, undoit);
                 flags.docalc = true;
             }
-            // all the other ones, just modify that one attribute
-            else param.set(newVal);
+            else {
+                var moduleAttrs = (contFull._module || {}).attributes || {};
+                var valObject = Lib.nestedProperty(moduleAttrs, ai).get() || {};
 
+                if(valObject.arrayOk && (Array.isArray(newVal) || Array.isArray(oldVal))) {
+                    flags.docalc = true;
```

Comment:
@alexcjohnson using `_module.attributes` here to find the `arrayOk` keys worked surprisingly well. We could declare a few more keys in the attributes like to resolve https://github.com/plotly/plotly.js/issues/648 at some point.